File: problems/Binary Search Tree Iterator/bstIterator.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class BSTIterator {

    public BSTIterator(TreeNode root) {
        leftMostInorder(root);
    }

    /** @return the next smallest number */
    public int Next() {
        var nextNode = stack.Pop();

        if (null != nextNode.right) {
            leftMostInorder(nextNode.right);
        }

        return nextNode.val;
    }

    /** @return whether we have a next smallest number */
    public bool HasNext() {
        return 0 < stack.Count;
    }

    private readonly Stack<TreeNode> stack = new Stack<TreeNode>();
    private void leftMostInorder(TreeNode node) {
        while (null != node) {
            stack.Push(node);
            node = node.left;
        }
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.Next();
 * bool param_2 = obj.HasNext();
 */

File: problems/Binary Search Tree Iterator/bstIterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {

    public BSTIterator(TreeNode root) {
        idx = 0;
        inorder = new ArrayList<>();

        fillInOrder(root, inorder);
    }

    public int next() {
        return inorder.get(idx++);
    }

    public boolean hasNext() {
        return idx < inorder.size();
    }

    private int idx;
    private List<Integer> inorder;

    private void fillInOrder(TreeNode node, List<Integer> inorder) {
        if (null == node) {
            return;
        }

        fillInOrder(node.left, inorder);
        inorder.add(node.val);
        fillInOrder(node.right, inorder);
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */
